### PR TITLE
Phase1: Add configuration objects and update tests

### PR DESCRIPTION
**Pull request by nagi1**
In this PR I 
- Added opinionated configuration object you may review names of the methods/properties.
- Dropped support for php 7 and laravel 7.
- Add return types to some of the functions that didn't have return types.

- Weren't able to move enable/disable logging into the config object.
- and recorded event array as well.

- Please review the change to proceed to Pipeline implementation !
- need advice on updating the docs.

### CODE REVIEW

File: tests/ActivityLoggerBatchTest.php [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+
+        $this->assertEquals($firstUuid, $secondUuid);
+    }
+
+
+    /** @test */
+    public function it_will_not_generate_uuid_if_end_batch_before_starting()
+    {
+        app(ActivityLoggerBatch::class)->endBatch();
```
Comment (Gummibeer) [inline tests/ActivityLoggerBatchTest.php:74]: This end and also the start call aren't important for this test

File: tests/ActivityLoggerBatchTest.php [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+        $this->assertNotNull($firstUuid);
+        $this->assertNull($nullUuid);
+
+        $this->assertNotSame($firstUuid, $nullUuid);
+    }
+
+
+    /** @test */
+    public function batch_stress_test()
```
Comment (Gummibeer) [inline tests/ActivityLoggerBatchTest.php:122]: # **Dead Or Alive** - _You Spin Me Round_

> You spin me right round, baby
> Right round like a record, baby
> Right round round round
> You spin me right round, baby
> Right round like a record, baby
> Right round round round

https://www.youtube.com/watch?v=PGNiXGX2nLU

**Review by Gummibeer — CHANGES_REQUESTED** [review]
Awesome outstanding heroic work here!  🤯😱🤩
🏆🥇

I don't know why - but could you merge in the latest `v4` branch to resolve the conflicts?
After that I will have a look again and also check it out on my local machine.
So far definitely amazing work and refactoring! 🚀

File: tests/ActivityLoggerBatchTest.php (hunk 2 above)
Comment (nagi1) [reply tests/ActivityLoggerBatchTest.php:122]: After making `ActivityLoggerBatch` class run for a 3 miles it totally should named like this ^^)

File: src/ActivitylogOptions.php [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    public array $logAttributes = [];
+
+    public array $ignoredAttributes = [];
+
+    public array $dontLogIfAttributesChangedBag = [];
+
+    public ?Closure $descriptionForEvent;
+
+    public static function create(): self
```
[the remark this answers is not in the document]
Comment (nagi1) [reply src/ActivitylogOptions.php:27]: I borrowed this naming from spatie/sluggable configuration object. I dont know if it'll make any difference, but laravel associate create and make with database related things while we are clearly not in the same context.

what do you think?

File: src/ActivitylogOptions.php [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    {
+        return new static;
+    }
+
+    public static function defaults(): self
+    {
+        return static::create();
+    }
+
+
```
[the remark this answers is not in the document]
Comment (nagi1) [reply src/ActivitylogOptions.php:37]: Laravel has some "convenience" methods that do exactly the same as the underplaying classes/methods like many of the helpers it provides.

I added this method when I encountered some tests that only create `Article` model and didnt add any "specific" configuration that means that class will use the defaults options provided by the package.

So I thought using `defaults()` would be more clear than `make()` or `create()`.
It's opinions after all ^^)

[the remark this answers is not in the document]
Comment (Gummibeer) [reply src/ActivitylogOptions.php:37]: How about dropping the `make` in that case? 🤔
We always start an option object with defaults we think are okay. So even if you customize it you only adjust single options on the defaults.

File: src/ActivitylogOptions.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gummibeer) [reply src/ActivitylogOptions.php:27]: If we keep this method I prefer `make()` - but see my comment on the `defaults()` method.

**Comment by nagi1** [comment]
> Awesome outstanding heroic work here! 
> 
> 
> I don't know why - but could you merge in the latest `v4` branch to resolve the conflicts?
> After that I will have a look again and also check it out on my local machine.
> So far definitely amazing work and refactoring!

These conflicts are nasty, plus I noticed that I just added the new `batch_uuid` column in the main migration file. so I need to get that fixed in new migration file too!

File: src/ActivitylogOptions.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (nagi1) [reply src/ActivitylogOptions.php:37]: Yeah I think this is even better, less is more in this case :)

File: src/ActivitylogOptions.php [hunk 5]
```diff
@@ -0,0 +1,129 @@
+<?php
+
+namespace Spatie\Activitylog;
+
+use Closure;
+
+class ActivitylogOptions
```
[the remark this answers is not in the document]
Comment (nagi1) [reply src/ActivitylogOptions.php:7]: What do you mean?

File: tests/ActivityLoggerBatchTest.php [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    {
+        app(ActivityLoggerBatch::class)->startBatch();
+
+        $firstUuid = app(ActivityLoggerBatch::class)->getUuid();
+
+        app(ActivityLoggerBatch::class)->startBatch();
+
+        $secondUuid = app(ActivityLoggerBatch::class)->getUuid();
+
+        app(ActivityLoggerBatch::class)->endBatch();
```
[the remark this answers is not in the document]
Comment (nagi1) [reply tests/ActivityLoggerBatchTest.php:64]: Please expand on this, what's need to be done here?

File: tests/ActivityLoggerBatchTest.php (hunk 1 above)
Comment (nagi1) [reply tests/ActivityLoggerBatchTest.php:74]: The test says end before starts, and that's exactly what It does.

**Comment by nagi1** [comment]
Moved to v4 #866

File: tests/ActivityLoggerBatchTest.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (Gummibeer) [reply tests/ActivityLoggerBatchTest.php:64]: I would reduce tests to the absolute minimum. As you don't assert anything on the batcher after this end call it's not needed. Same for several other testcases.
Or you fully close all and assert UUID null or is open false.

**Pull request closed without merging**